src/RA_httpthread.cpp

```diff
@@ -252,31 +247,74 @@ void RAWeb::SetUserAgentString()
 #ifndef NTSTATUS
     #define NTSTATUS long
 #endif
-    using fnRtlGetVersion = NTSTATUS(NTAPI*)(PRTL_OSVERSIONINFOEXW lpVersionInformation);
-    auto RtlGetVersion = (fnRtlGetVersion)GetProcAddress(GetModuleHandleW(L"ntdll.dll"), "RtlGetVersion");
-    if (RtlGetVersion)
     {
-        RTL_OSVERSIONINFOEXW osVersion = {};
+        // needs to be XP or higher no point in continuing
+        // Windows 7 doesn't have the version helper api....
+        // TBD: Consider Service Pack?
+        {
+            auto IsXPOrGreater =[]() noexcept
+            {
+                OSVERSIONINFOEXW osvi{
+                    sizeof(osvi), HIBYTE(_WIN32_WINNT_WINXP) , LOBYTE(_WIN32_WINNT_WINXP)
+                };
+                const auto dwlConditionMask
+                {
+                    VerSetConditionMask(VerSetConditionMask(VerSetConditionMask(
+                        0ULL, VER_MAJORVERSION, VER_GREATER_EQUAL),
+                        VER_MINORVERSION, VER_GREATER_EQUAL),
+                        VER_SERVICEPACKMAJOR, VER_GREATER_EQUAL)
+                };
+
+                return VerifyVersionInfoW(&osvi, VER_MAJORVERSION | VER_MINORVERSION |
+                                          VER_SERVICEPACKMAJOR, dwlConditionMask) != FALSE;
+            };
+            if (!IsXPOrGreater())
+                return;
+        }
+
+        using fnRtlGetVersion = NTSTATUS(NTAPI*)(PRTL_OSVERSIONINFOEXW lpVersionInformation);
+
+        const auto& ntModule{ ::GetModuleHandleW(L"ntdll.dll") };
+        if (!ntModule)
+            return;
+
+        RTL_OSVERSIONINFOEXW osVersion{};
+        const auto& RtlGetVersion{
+            reinterpret_cast<fnRtlGetVersion>(::GetProcAddress(ntModule, "RtlGetVersion"))
+        };
+        if (!RtlGetVersion)
+            return;
         RtlGetVersion(&osVersion);
-        if (osVersion.dwMajorVersion > 0)
+        
         {
-            sUserAgent.append("WindowsNT ");
-            sUserAgent.append(std::to_string(osVersion.dwMajorVersion));
-            sUserAgent.append(".");
-            sUserAgent.append(std::to_string(osVersion.dwMinorVersion));
+            std::ostringstream oss;
+            oss << "WindowsNT " << osVersion.dwMajorVersion << '.' << osVersion.dwMinorVersion
+                << '.' << osVersion.dwBuildNumber;
+            sUserAgent.append(oss.str());
         }
     }
 
     sUserAgent.append(") Integration/");
+    {
+        std::ostringstream oss;
+        // Compile-time string concatenation will need some user-defined stuff/work.
+        oss << "WindowsNT " << RA_INTEGRATION_VERSION_MAJOR << '.' << RA_INTEGRATION_VERSION_MINOR
+            << '.' << RA_INTEGRATION_VERSION_REVISION << '.' << RA_INTEGRATION_VERSION_MODIFIED;
+        sUserAgent.append(oss.str());
+    }
```

Comment:
Instead of using `sprintf` once, you're relying on the `<<` operator to call `std::to_string` four times and using an `ostringstream` to build an intermediate string. How is this better?

Reply:
I need to check again but I thought the macros were numbers except the last one. It was for speed. It go back but using a static array instead

Reply:
I don't know what code you're talking about but `operator<<` does not call `to_string` it uses `ostream`. So I still think this is better. I will run some benchmarks just to be sure. Theoretically this should be faster.

Reply:
And theory was disproved, implementations are not made equal I guess. It was faster last time I bench-marked using GCC (for research not this project).